proposals/0000-transient-storages.md

```diff
@@ -0,0 +1,336 @@
+---
+Title: transient storage host functions
+Number: 0
+Status: Proposed
+Authors:
+  - cheme
+Created: 2023-04-xx
+Category: Runtime Environment
+Requires:
+Replaces:
+--- 
+
+
+Current Implementation: https://github.com/cheme/substrate/tree/transient
+
+
+## Summary
+
+This PPPs defines new host functions to keep trace of transient data accross a block processing.
+
+It is derived and modified from initial writing https://github.com/paritytech/substrate/issues/12577, credits to Gavin Wood.
+
+It exposes two new sets of host functions, blob storage an api around bytes storage (similar to rust  `Vec<u8>`), and ordered key values storage (similar to rust `BTreeMap<Vec<u8>, Vec<u8>>` storage).
+
+Each storage structure are defined and address by a name with no limit in number except the block weight.
+Each structure can be hashed or reduced to a single merkle root to possible store it on state.
+Each structure can define if it should be send to client storage.
+
+## Motivation
+
+From https://github.com/paritytech/substrate/issues/5396: 
+"Right now we abuse storage for intra-block data such as block number, parent hash and block author as well as various housekeeping information and flags like whether we set the uncles Authorship::DidSetUncles.
+
+When initially writing, this incurs an extra trie lookup, which is slow. Instead there should be another host API, which works exactly like set_storage/get_storage but has no trie backing, so it never tries to lookup the value in the trie, nor does it write the value at the end of the block." Credits Gavin Wood.
+
+Additionally content such as events often are used in a log based manner (append only) with possibly a larger size than usual content.
+
+Hashing through host function involves passing all data at once, and is not memory efficient.
+
+## Implementation
+
+Transient storage act as current state storage, but without a persistent backend.
+
+This implies that the storage must support commiting and reverting transaction with `ext_storage_commit_transaction` or `ext_storage_rollback_transaction`.
+This transactional support is both at transient storage content and at transient storage definition (a delete transient storage will be restore on rollback).
+
+Btree and blob are using a specific `Mode`, either `drop` or `archive` passed respectively as the byte 0 or 1. When using `drop` the data will not be send from the runtime executor to the calling client. When using `archive` the committed state of the transient storage will be passed as a change set to the client calling runtime executor.
```

Comment:
> When using `archive` the committed state of the transient storage will be passed as a change set to the client calling runtime executor.

Isn't that an implementation detail that isn't relevant here?

Reply:
Yes we could stick to "archive indicate that information should be storable and fetch-able from the client".